File: src/C++/load_imp_par.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_set>
#include <chrono>
#include <future>
#include <pqxx/pqxx>
#include <iomanip>
#include "csv.h"

const std::string DB_CONFIG = "host=localhost dbname=project1 user=postgres password=123 port=5432";

// 自定义元组哈希
namespace std {
    template<typename... T>
    struct hash<tuple<T...>> {
        size_t operator()(const tuple<T...>& key) const {
            size_t seed = 0;
            apply([&seed](const T&... args) {
                ((seed ^= hash<T>{}(args) + 0x9e3779b9 + (seed << 6) + (seed >> 2)), ...);
            }, key);
            return seed;
        }
    };
}

// 数据清洗函数
std::string clean_date(const std::string& date_str) {
    std::string cleaned = date_str;
    cleaned.erase(0, cleaned.find_first_not_of(" \t"));
    cleaned.erase(cleaned.find_last_not_of(" \t") + 1);

    if (cleaned.empty() || cleaned == "nan") 
        return "1970-01-01";

    struct tm tm = {};
    const char* formats[] = {
        "%Y/%m/%d", "%Y%m%d", "%Y-%m-%d", 
        "%d/%m/%Y", "%m/%d/%Y", "%d-%m-%Y"
    };

    for (const auto& fmt : formats) {
        if (strptime(cleaned.c_str(), fmt, &tm)) {
            char buffer[11];
            strftime(buffer, sizeof(buffer), "%Y-%m-%d", &tm);
            return buffer;
        }
    }
    return "1970-01-01";
}

// 并行插入模板
template <typename T>
void parallel_table_insert(const std::string& table_name, 
                            std::initializer_list<std::string_view> columns,
                          const std::unordered_set<T>& dataset) {
    try {
        pqxx::connection conn(DB_CONFIG);
        pqxx::work txn(conn);
        
        pqxx::table_path tablePath{table_name};
        auto stream = pqxx::stream_to::table(txn, tablePath, columns);
        
        for (const auto& item : dataset) {
            std::apply([&](const auto&... args) {
                stream.write_values(args...);
            }, item);
        }
        
        stream.complete();
        txn.commit();
        std::cout << "成功插入 " << table_name << " (" << dataset.size() << " 行)\n";
    } catch (const std::exception& e) {
        std::cerr << "插入 " << table_name << " 失败: " << e.what() << std::endl;
        throw;
    }
}

int main() {
    auto total_start = std::chrono::high_resolution_clock::now();
    
    try {
        // 阶段1: 数据加载
        auto load_start = std::chrono::high_resolution_clock::now();
        
        io::CSVReader<20, io::trim_chars<>, io::double_quote_escape<',','\"'>> in("/home/wgx/database/CS307-project1/resources/output25S.csv");
        in.read_header(io::ignore_extra_column, 
            "contract number", "client enterprise", "supply center", "country", "city", 
            "industry", "product code", "product name", "product model", "unit price", 
            "quantity", "contract date", "estimated delivery date", "lodgement date", 
            "director", "salesman", "salesman number", "gender", "age", "mobile phone");

        std::unordered_set<std::tuple<std::string, std::string>> supply_centers;
        std::unordered_set<std::tuple<std::string, std::string, std::string, std::string, std::string>> clients;
        std::unordered_set<std::tuple<std::string, std::string>> products;
        std::unordered_set<std::tuple<std::string, std::string, int>> product_models;
        std::unordered_set<std::tuple<int, std::string, std::string, std::string, int>> sales;
        std::unordered_set<std::tuple<std::string, std::string, std::string>> contracts;
        std::vector<std::tuple<std::string, std::string, std::string, int, std::string, std::string, int>> orders;

        std::string contract_num, client_ent, supply_ctr, country, city, industry, 
                    product_cd, product_nm, product_mdl, contract_dt, est_deliv_dt, lodge_dt,
                    director, salesman, gender, mobile;
        int unit_price, quantity, salesman_num, age;

        while (in.read_row(contract_num, client_ent, supply_ctr, country, city, industry,
                          product_cd, product_nm, product_mdl, unit_price, quantity,
                          contract_dt, est_deliv_dt, lodge_dt, director, salesman,
                          salesman_num, gender, age, mobile)) {
            
            // 清洗日期字段
            contract_dt = clean_date(contract_dt);
            est_deliv_dt = clean_date(est_deliv_dt);
            lodge_dt = clean_date(lodge_dt);

            // 收集数据
            supply_centers.emplace(supply_ctr, director);
            clients.emplace(client_ent, country, supply_ctr, city, industry);
            products.emplace(product_cd, product_nm);
            product_models.emplace(product_cd, product_mdl, unit_price);
            sales.emplace(salesman_num, salesman, gender, mobile, age);
            contracts.emplace(contract_num, client_ent, contract_dt);
            orders.emplace_back(contract_num, product_cd, product_mdl, quantity, 
                              est_deliv_dt, lodge_dt, salesman_num);
        }

        auto load_end = std::chrono::high_resolution_clock::now();
        std::cout << "数据加载完成 | 耗时: " 
                << std::chrono::duration<double>(load_end - load_start).count() 
                << "s\n";

        // 阶段2: 并行插入
        std::vector<std::future<void>> tasks;
        
        // 第一组并行任务（无依赖表）
        tasks.push_back(std::async(std::launch::async, [&](){
            parallel_table_insert<decltype(supply_centers)::value_type>( 
            "supply_center", 
            {"center_name", "director"}, 
            std::ref(supply_centers));}
        ));

        tasks.push_back(std::async(std::launch::async, [&](){
            parallel_table_insert<decltype(products)::value_type>(
            "product", 
            {"product_code", "product_name"}, 
            std::ref(products));}
        ));

        // 等待第一组完成
        for (auto& t : tasks) t.wait();
        tasks.clear();

        // 第二组并行任务（部分依赖）
        tasks.push_back(std::async(std::launch::async, [&](){
            parallel_table_insert<decltype(clients)::value_type>(
            "client", 
            {"client_name", "country", "supply_center", "city", "industry"}, 
            std::ref(clients));}
        ));

        tasks.push_back(std::async(std::launch::async, [&](){
            parallel_table_insert<decltype(product_models)::value_type>(
            "product_model", 
            {"product_code", "product_model", "unit_price"}, 
            std::ref(product_models));}
        ));

        tasks.push_back(std::async(std::launch::async, [&](){
            parallel_table_insert<decltype(sales)::value_type>(
            "sales", 
            {"salesman_number", "salesman_name", "gender", "mobile_number", "age"}, 
            std::ref(sales));}
        ));

        // 等待第二组完成
        for (auto& t : tasks) t.wait();
        tasks.clear();

        // 第三阶段（顺序执行）
        parallel_table_insert("contract", 
            {"contract_number", "client_name", "contract_date"}, 
            contracts);

        // 插入订单详情
        if (!orders.empty()) {
            pqxx::connection conn(DB_CONFIG);
            pqxx::work txn(conn);
            
            auto stream = pqxx::stream_to::table(txn, pqxx::table_path{"order_detail"}, 
                {"contract_number", "product_code", "product_model", "quantity", 
                 "estimated_delivery_date", "lodgement_date", "salesman_number"});
            
            for (const auto& o : orders) {
                std::apply([&](const auto&... args) {
                    stream.write_values(args...);
                }, o);
            }
            
            stream.complete();
            txn.commit();
            std::cout << "成功插入 order_detail (" << orders.size() << " 行)\n";
        }

    } catch (const std::exception& e) {
        std::cerr << "主程序异常: " << e.what() << std::endl;
        return 1;
    }

    auto total_end = std::chrono::high_resolution_clock::now();
    std::cout << "\n总运行时间: "
            << std::chrono::duration<double>(total_end - total_start).count() 
            << " 秒\n";

    return 0;
}

File: src/Python/load_imp_par.py
import asyncio
import asyncpg
import pandas as pd
import numpy as np
from datetime import datetime, date
import os
from typing import List, Tuple, Optional

# 数据库连接池配置
DB_CONFIG = {
    "host": "localhost",
    "database": "project1",
    "user": "normaluser",
    "password": "123",
    "port": 5432,
    "min_size": 5,
    "max_size": 10,
    "command_timeout": 60
}

file_path = "../../resources/output25S.csv"
abs_file_path = os.path.abspath(file_path)

def clean_date(date_str: str) -> Optional[date]:
    """增强型日期清洗函数"""
    if pd.isna(date_str):
        return None
    
    date_str = str(date_str).strip()
    formats = [
        "%Y/%m/%d", "%Y%m%d", 
        "%Y-%m-%d", "%d/%m/%Y",
        "%m/%d/%Y", "%d-%m-%Y"
    ]
    
    for fmt in formats:
        try:
            dt = datetime.strptime(date_str.split(' ')[0], fmt)
            return dt.date()
        except ValueError:
            continue
    
    print(f"无法解析的日期格式: {date_str}")
    return None

async def create_connection_pool() -> asyncpg.pool.Pool:
    """创建数据库连接池"""
    return await asyncpg.create_pool(**DB_CONFIG)

async def copy_with_conflict_handling(
    conn: asyncpg.Connection,
    table: str,
    columns: List[str],
    data: List[tuple],
    conflict_columns: List[str] = None
) -> int:
    """安全批量插入"""
    if not data:
        return 0
    
    temp_table = f"temp_{table}"
    
    async with conn.transaction():
        # 创建临时表
        await conn.execute(f"""
            CREATE TEMP TABLE {temp_table} 
            (LIKE {table} INCLUDING DEFAULTS)
            ON COMMIT DROP
        """)
        
        # 执行COPY
        await conn.copy_records_to_table(temp_table, records=data, columns=columns)
        
        # 构建插入语句
        insert_sql = f"""
            INSERT INTO {table} 
            SELECT * FROM {temp_table}
            {f'ON CONFLICT ({", ".join(conflict_columns)}) DO NOTHING' 
                if conflict_columns else ''}
        """
        
        result = await conn.execute(insert_sql)
        return int(result.split()[-1])

async def parallel_data_import(
    pool: asyncpg.pool.Pool,
    tasks: List[Tuple[str, List[str], List[tuple], List[str]]]
) -> None:
    """增强型并行导入"""
    async def worker(task):
        try:
            async with pool.acquire() as conn:
                table, cols, data, conflict = task
                start = datetime.now()
                affected = await copy_with_conflict_handling(conn, table, cols, data, conflict)
                print(f"{table.ljust(15)} 插入 {affected} 行 | 耗时: {datetime.now()-start}")
        except Exception as e:
            print(f"\n处理任务 {table} 时出错: {str(e)}")
            if data:
                print(f"首条问题数据示例: {data[0]}")
            raise

    # 分批次处理避免内存溢出
    batch_size = 50000
    all_tasks = []
    for task in tasks:
        table, cols, data, conflict = task
        for i in range(0, len(data), batch_size):
            batched_data = data[i:i+batch_size]
            all_tasks.append((table, cols, batched_data, conflict))
    
    # 并行执行
    await asyncio.gather(*[worker(t) for t in all_tasks])

async def main():
    """主函数"""
    total_start = datetime.now()
    print(f"{'='*40}\n数据导入开始 @ {total_start}\n{'='*40}")
    
    # 检查文件
    if not os.path.exists(abs_file_path):
        raise FileNotFoundError(f"CSV文件不存在: {abs_file_path}")
    
    # 阶段1：数据加载和清洗
    print("\n[阶段1] 数据读取和预处理...")
    start_load = datetime.now()
    
    # 明确指定数据类型
    dtype_spec = {
        'salesman number': 'str',
        'mobile phone': 'str',
        'contract number': 'str',
        'product code': 'str',
        'product model': 'str',
        'quantity': 'Int64',
        'unit price': 'float64'
    }
    
    df = pd.read_csv(abs_file_path, dtype=dtype_spec)
    
    # 清洗关键字段
    numeric_cols = ['quantity', 'unit price']
    for col in numeric_cols:
        df[col] = pd.to_numeric(df[col], errors='coerce')
        df[col] = df[col].fillna(0) if 'price' in col else df[col].fillna(0).astype(int)
    
    # 日期字段处理
    date_cols = ['contract date', 'estimated delivery date', 'lodgement date']
    for col in date_cols:
        df[col] = df[col].apply(clean_date)
    
    # 处理文本字段的NaN
    text_cols = ['client enterprise', 'country', 'supply center', 'city', 'industry']
    df[text_cols] = df[text_cols].fillna('').astype(str)
    
    print(f"数据加载完成 | 总记录数: {len(df):,} | 耗时: {datetime.now()-start_load}")
    
    # 阶段2：数据准备
    print("\n[阶段2] 数据准备...")
    start_prepare = datetime.now()
    
    # 确保salesman number转换为整数
    df['salesman number'] = pd.to_numeric(
        df['salesman number'], 
        errors='coerce'
    ).fillna(0).astype(int)
    
    # 构建数据集
    supply_data = df[['supply center', 'director']].drop_duplicates().to_numpy().tolist()
    client_data = df[['client enterprise', 'country', 'supply center', 'city', 'industry']].drop_duplicates().to_numpy().tolist()
    product_data = df[['product code', 'product name']].drop_duplicates().to_numpy().tolist()
    model_data = df[['product code', 'product model', 'unit price']].drop_duplicates().to_numpy().tolist()
    sales_data = df[['salesman number', 'salesman', 'gender', 'mobile phone', 'age']].drop_duplicates().to_numpy().tolist()
    contract_data = df[['contract number', 'client enterprise', 'contract date']].drop_duplicates().to_numpy().tolist()
    
    # 明确转换order数据类型
    order_data = []
    for _, row in df.iterrows():
        order_data.append((
            str(row['contract number']),
            str(row['product code']),
            str(row['product model']),
            int(row['quantity']),
            row['estimated delivery date'],
            row['lodgement date'],
            int(row['salesman number'])
        ))
    
    print(f"数据准备完成 | 耗时: {datetime.now()-start_prepare}")
    
    # 阶段3：并行导入
    print("\n[阶段3] 并行数据导入...")
    pool = await create_connection_pool()
    
    try:
        # 分阶段任务配置
        stage_config = [
            {
                'name': '基础表',
                'tasks': [
                    ('supply_center', ['center_name', 'director'], supply_data, ['center_name']),
                    ('product', ['product_code', 'product_name'], product_data, ['product_code']),
                    ('sales', ['salesman_number', 'salesman_name', 'gender', 'mobile_number'], sales_data, ['salesman_number'])
                ]
            },
            {
                'name': '关联表',
                'tasks': [
                    ('client', ['client_name', 'country', 'supply_center', 'city', 'industry'], client_data, ['client_name']),
                    ('product_model', ['product_code', 'product_model', 'unit_price'], model_data, ['product_code', 'product_model'])
                ]
            },
            {
                'name': '合同数据',
                'tasks': [
                    ('contract', ['contract_number', 'client_name', 'contract_date'], contract_data, ['contract_number'])
                ]
            },
            {
                'name': '订单详情',
                'tasks': [
                    ('order_detail', [
                        'contract_number', 'product_code', 'product_model', 
                        'quantity', 'estimated_delivery_date', 'lodgement_date', 'salesman_number'
                    ], order_data, None)
                ]
            }
        ]
        
        # 分阶段执行
        total_import_start = datetime.now()
        for stage in stage_config:
            stage_start = datetime.now()
            print(f"\n--- 正在导入 {stage['name']} ---")
            await parallel_data_import(pool, stage['tasks'])
            print(f"阶段完成 | 耗时: {datetime.now()-stage_start}")
        
        print(f"\n所有数据导入完成 | 总导入耗时: {datetime.now()-total_import_start}")
        
    except Exception as e:
        print(f"\n致命错误: {type(e).__name__}: {str(e)}")
    finally:
        await pool.close()
    
    total_end = datetime.now()
    print(f"\n{'='*40}\n数据导入结束 @ {total_end}")
    print(f"总运行时间: {total_end - total_start}\n{'='*40}")

if __name__ == "__main__":
    asyncio.run(main())